### DISCUSSION TITLE
Navier-Stokes: computing 3D velocity gradient

### DISCUSSION THREAD

**Discussion by Watchmojo** [mfem/mfem#2633] [General]
I am interested to add turbulence models in the NS solver. As the first step, the velocity gradient tensor needs to be calculated. I did the following:

Step 1: added `std::vector<ParGridFunction> du_gf` in the naver_solver class, and initialized using 
            `du_gf.resize(vfes->GetVDim()); `
           `for (auto it = du_gf.begin(); it != du_gf.end(); it++)`
             `it->SetSpace(vfes);`

Step 2: "Hijacked" ComputeCurl3D and made it a new function like this:
`void NavierSolver::ComputeGrad3D(ParGridFunction &u, std::vector<ParGridFunction> &du)`

The navier_solver.hpp navier_solver.cpp are attached here. 
-----
1. Would you please comment? 
2. Is the vector du at the end of this function in the physical or function space? 

[navier_solver.zip](https://github.com/mfem/mfem/files/7464969/navier_solver.zip)

**Comment by jandrej** [comment]
I won't review the attached file.

To compute the gradient of a vector valued `GridFunction` wouldn't the method
```
   void GetVectorGradient(ElementTransformation &tr, DenseMatrix &grad) const;
```
suffice?

**Comment by suqingji** [comment]
Hi, Dear prof.
Have you finished turbulence model in ns solver.
thanks.
